docs/source/federation/implementing.md

```diff
@@ -200,4 +200,66 @@ server.listen().then(({ url }) => {
 
 By leveraging the `buildService` function, we're able to customize how requests are sent to our federated services. In this example, we return a custom `RemoteGraphQLDataSource`. The datasource allows us to modify the outgoing request with information from the Apollo Server `context` before it's sent. Here, we're adding the `user-id` header to pass an authenticated user id to downstream services.
 
+In a similar vein, the `didReceiveResponse` hook allows us to inspect a service's `response` in order to modify the `context`. The lifecycle of a request to a federated server involves a number of responses, which may contain headers that need to be passed back to the client. Suppose our services use the <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control" target="_blank" stop="_italicizing" rel="noopener noreferrer">Cache-Control</a> header's `max-age` value. We may have receive a number of these values from our services but we can only respond with one, which should be the minimum of all of the values. We can implement this using the `didReceiveResponse` hook and an `ApolloServerPlugin`:
+
+```javascript{4-20,48,59-71}
+const { ApolloServer } = require('apollo-server');
+const { ApolloGateway, RemoteGraphQLDataSource } = require('@apollo/gateway');
+
+class DataSourceWithCacheControl extends RemoteGraphQLDataSource {
+  async didReceiveResponse(response, request, context) {
+    const body = await super.didReceiveResponse(response, request, context);
+    // Parse the Cache-Control header and update the value on context if it's
+    // _less_ than the current value (or if no value is currently set).
+    const cacheControl = response.headers.get('Cache-Control');
+    if (cacheControl) {
+      const result = cacheControl.match(/max-age=(\d*)/);
+      if (Array.isArray(result) && result[1]) {
+        context.cacheControl.maxAge = context.cacheControl.maxAge
+          ? Math.min(context.cacheControl.maxAge, Number(result[1]))
+          : Number(result[1]);
+      }
+    }
+    return body;
```

Comment:
I do like this example and think it's much better than surrogate keys.  However, I think we should just tone this town to be a generic example representing a number (e.g.).

More technically, because cache control is "complicated" and there is extraordinary variation ([_`Vary`'d_ literally, sometimes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary)!) in how such a header is interpreted (plus, `public`/`private` rules, etc.).  For example, if the downstream service isn't present at all, it could be argued that that represents a `max-age` of `0`, which wouldn't work in the above example.  Ultimately, I worry that this would be an endorsement for its use.

But just-for-fun editorial on this suggestion:

Of Note:

* `\d*` should be `\d+` as there _must_ be at least one digit.  (`*` means _0 or more_, whereas `+` is _1 or more_).
* I default to `0` when `Cache-control` is absent.
* Shorter, but I think same result.

```js
class DataSourceWithCacheControl extends RemoteGraphQLDataSource {
  async didReceiveResponse(response, request, context) {
    // Parse the Cache-Control header and update the value on context if it's
    // _less_ than the current value (or if no value is currently set).
    const [ , seconds = 0] = (response.headers.get('Cache-Control') || "").match(/max-age=(\d+)/) || [];
    context.cacheControl.maxAge = Math.min(context.cacheControl.maxAge || +seconds, +seconds);
    // Important: Invoke the super method to return the response body.
    return await super.didReceiveResponse(response, request, context);
  }
}
```

But again, I don't think we necessarily want a `cache-control` example because of the overall technical complications of interpreting it properly with such a small amount of code.

Reply:
Resolved in 8898ad41e98722b3960b029984506a15b2336238